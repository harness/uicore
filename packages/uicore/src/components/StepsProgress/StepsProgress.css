/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.stepsProgress {
  .progressBar {
    margin-bottom: var(--spacing-xlarge);
    .progressHeight {
      height: 10px;
      border-radius: 0;
      background: transparent;
      box-shadow: 0px 0px 2px 1px inset var(--grey-300);
    }
  }
  .stepWrap {
    display: flex;
    .stepLabelContainer {
      flex: 1;
    }
  }
}
