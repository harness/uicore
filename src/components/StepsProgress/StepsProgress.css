.stepsProgress {
  .progressBar {
    margin-bottom: var(--spacing-xlarge);
    .progressHeight {
      height: 10px;
      border-radius: 0;
      background: transparent;
      box-shadow: 0px 0px 2px 1px inset var(--grey-300);
    }
  }
  .stepWrap {
    display: flex;
    .stepLabelContainer {
      flex: 1;
    }
  }
}
