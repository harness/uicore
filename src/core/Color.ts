export const Color = {
  PRIMARY_BG: 'primaryBG',
  PRIMARY_10: 'primary10',
  PRIMARY_9: 'primary9',
  PRIMARY_8: 'primary8',
  PRIMARY_7: 'primary7',
  PRIMARY_6: 'primary6',
  PRIMARY_5: 'primary5',
  PRIMARY_4: 'primary4',
  PRIMARY_3: 'primary3',
  PRIMARY_2: 'primary2',
  PRIMARY_1: 'primary1',
  GREY_1000: 'grey1000',
  GREY_900: 'grey900',
  GREY_800: 'grey800',
  GREY_700: 'grey700',
  GREY_600: 'grey600',
  GREY_500: 'grey500',
  GREY_400: 'grey400',
  GREY_300: 'grey300',
  GREY_200: 'grey200',
  GREY_100: 'grey100',
  GREY_50: 'grey50',
  GREY_0: 'grey0',
  GREEN_900: 'green900',
  GREEN_800: 'green800',
  GREEN_700: 'green700',
  GREEN_600: 'green600',
  GREEN_500: 'green500',
  GREEN_400: 'green400',
  GREEN_300: 'green300',
  GREEN_200: 'green200',
  GREEN_100: 'green100',
  GREEN_50: 'green50',
  BLUE_900: 'blue900',
  BLUE_800: 'blue800',
  BLUE_700: 'blue700',
  BLUE_600: 'blue600',
  BLUE_500: 'blue500',
  BLUE_400: 'blue400',
  BLUE_300: 'blue300',
  BLUE_200: 'blue200',
  BLUE_100: 'blue100',
  BLUE_50: 'blue50',
  YELLOW_900: 'yellow900',
  YELLOW_800: 'yellow800',
  YELLOW_700: 'yellow700',
  YELLOW_600: 'yellow600',
  YELLOW_500: 'yellow500',
  YELLOW_400: 'yellow400',
  YELLOW_300: 'yellow300',
  YELLOW_200: 'yellow200',
  YELLOW_100: 'yellow100',
  YELLOW_50: 'yellow50',
  ORANGE_900: 'orange900',
  ORANGE_800: 'orange800',
  ORANGE_700: 'orange700',
  ORANGE_600: 'orange600',
  ORANGE_500: 'orange500',
  ORANGE_400: 'orange400',
  ORANGE_300: 'orange300',
  ORANGE_200: 'orange200',
  ORANGE_100: 'orange100',
  ORANGE_50: 'orange50',
  RED_900: 'red900',
  RED_800: 'red800',
  RED_700: 'red700',
  RED_600: 'red600',
  RED_500: 'red500',
  RED_400: 'red400',
  RED_300: 'red300',
  RED_200: 'red200',
  RED_100: 'red100',
  RED_50: 'red50',
  TEAL_900: 'teal900',
  TEAL_800: 'teal800',
  TEAL_700: 'teal700',
  TEAL_600: 'teal600',
  TEAL_500: 'teal500',
  TEAL_400: 'teal400',
  TEAL_300: 'teal300',
  TEAL_200: 'teal200',
  TEAL_100: 'teal100',
  TEAL_50: 'teal50',
  LIME_900: 'lime900',
  LIME_800: 'lime800',
  LIME_700: 'lime700',
  LIME_600: 'lime600',
  LIME_500: 'lime500',
  LIME_400: 'lime400',
  LIME_300: 'lime300',
  LIME_200: 'lime200',
  LIME_100: 'lime100',
  LIME_50: 'lime50',
  PURPLE_900: 'purple900',
  PURPLE_800: 'purple800',
  PURPLE_700: 'purple700',
  PURPLE_600: 'purple600',
  PURPLE_500: 'purple500',
  PURPLE_400: 'purple400',
  PURPLE_300: 'purple300',
  PURPLE_200: 'purple200',
  PURPLE_100: 'purple100',
  PURPLE_50: 'purple50',
  MAGENTA_900: 'magenta900',
  MAGENTA_800: 'magenta800',
  MAGENTA_700: 'magenta700',
  MAGENTA_600: 'magenta600',
  MAGENTA_500: 'magenta500',
  MAGENTA_400: 'magenta400',
  MAGENTA_300: 'magenta300',
  MAGENTA_200: 'magenta200',
  MAGENTA_100: 'magenta100',
  MAGENTA_50: 'magenta50',
  AQUA_500: 'aqua500',
  BLACK: 'black',
  BLACK_100: 'black100',
  WHITE: 'white',
  // Form styles
  FORM_BG: 'form',
  FORM_SECTION_BG: 'formSection',
  FORM_PANEL_BG: 'formPanel',
  FORM_SUBSECTION_BG: 'formSubSection',
  ERROR: 'error',
  WARNING: 'warning',
  SUCCESS: 'success',
  // Extra
  GREY_450: 'grey450',
  GREY_350: 'grey350',
  GREY_250: 'grey250',
  BLUE_450: 'blue450',
  BLUE_350: 'blue350',
  RED_450: 'red450',
  RED_350: 'red350',
  YELLOW_450: 'yellow450',
  YELLOW_350: 'yellow350',
  GREEN_450: 'green450',
  GREEN_350: 'green350',
  SEA_GREEN_500: 'seaGreen500'
}

export type Color = typeof Color[keyof typeof Color]
