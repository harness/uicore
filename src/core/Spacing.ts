// enum Spacing {
//   NONE = 'none',
//   XSMALL = 'xsmall',
//   SMALL = 'small',
//   MEDIUM= 'medium',
//   LARGE= 'large' ,
//   XLARGE= 'xlarge' ,
//   XXLARGE= 'xxlarge' ,
//   XXXLARGE= 'xxxlarge' ,
//   HUGE= 'huge'
// }

export type Spacing = 'none' | 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'xxlarge' | 'xxxlarge' | 'huge'
