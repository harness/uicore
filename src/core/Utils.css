.tooltipContainer {
  padding: var(--spacing-medium);
  max-width: '500px';
  max-height: '500px';
  overflow: 'auto';
  overflow-wrap: 'break-word';
  word-wrap: 'break-word';
  hyphens: 'auto';
}
