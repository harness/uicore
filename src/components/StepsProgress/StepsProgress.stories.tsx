/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary } from '@storybook/addon-docs/blocks'
import { Intent, StepsProgress } from '../..'
import { StepsProgressProps } from '../StepsProgress/StepsProgress'
import { omit } from 'lodash-es'

export default {
  title: 'Components / StepsProgress',

  component: StepsProgress,
  parameters: {
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>StepsProgress</Title>
            <Subtitle>
              <pre>
                <code>{`import {StepsProgress} from '@harness/uicore'`}</code>
              </pre>
            </Subtitle>

            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta
export const Basic: Story<StepsProgressProps> = args => {
  const {
    steps = [
      'Checking Delegate',
      'Establising Connection with the Delegate and Kubernetes',
      'Checking for heartbeat',
      'Verifying Connection'
    ],
    intent = Intent.SUCCESS,
    current = 1,
    currentStatus = 'PROCESS'
  } = args
  const argsCopy = omit(args, ['steps', 'intent', 'current', 'currentStatus'])
  return <StepsProgress steps={steps} intent={intent} current={current} currentStatus={currentStatus} {...argsCopy} />
}
export const Example1: Story<StepsProgressProps> = args => {
  const {
    steps = [
      'Checking Delegate',
      'Establising Connection with the Delegate and Kubernetes',
      'Checking for heartbeat',
      'Verifying Connection'
    ],
    intent = Intent.DANGER,
    current = 3,
    currentStatus = 'ERROR'
  } = args
  const argsCopy = omit(args, ['steps', 'intent', 'current', 'currentStatus'])
  return <StepsProgress steps={steps} intent={intent} current={current} currentStatus={currentStatus} {...argsCopy} />
}
export const Example2: Story<StepsProgressProps> = args => {
  const {
    steps = [
      'Checking Delegate',
      'Establising Connection with the Delegate and Kubernetes',
      'Checking for heartbeat',
      'Verifying Connection'
    ],
    intent = Intent.SUCCESS,
    current = 4,
    currentStatus = 'DONE'
  } = args
  const argsCopy = omit(args, ['steps', 'intent', 'current', 'currentStatus'])
  return <StepsProgress steps={steps} intent={intent} current={current} currentStatus={currentStatus} {...argsCopy} />
}
