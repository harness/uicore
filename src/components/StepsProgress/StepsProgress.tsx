import React from 'react'
import { Text } from '../Text/Text'
import { Container } from '../Container/Container'
import { Color } from '../../core/Color'
import css from './StepsProgress.css'
import { ProgressBar, IProgressBarProps, Intent } from '@blueprintjs/core'
import { Layout } from '../../layouts/Layout'
import { Icon } from '../../icons/Icon'

export interface StepsProgressProps {
  steps: Array<JSX.Element | string>
  current: number
  currentStatus: string
  intent?: Intent
}

interface StringMap {
  [key: string]: string
}

const fontsEnums: StringMap = {
  WAIT: 'normal',
  PROCESS: 'bold',
  DONE: 'bold',
  ERROR: ''
}

const statusEnums: StringMap = {
  WAIT: Color.YELLOW_500,
  PROCESS: Color.YELLOW_500,
  DONE: Color.GREEN_500,
  ERROR: Color.RED_500
}

const iconEnums: any = {
  WAIT: 'blank',
  PROCESS: 'steps-spinner',
  DONE: 'tick',
  ERROR: 'error'
}

export function StepsProgress(props: StepsProgressProps) {
  const { steps, current, currentStatus, intent } = props
  const stepSize = 1 / steps.length
  let calculatedStepSize = 0
  if (currentStatus === 'PROCESS') {
    calculatedStepSize = (current - 1) * stepSize
  } else {
    calculatedStepSize = current * stepSize
  }

  const progressBarProps: IProgressBarProps = {
    value: calculatedStepSize,
    stripes: false,
    intent: intent
  }
  let stepsWithStatus
  if (steps && steps.length) {
    stepsWithStatus = steps.map((data, index) => ({
      label: data,
      fonts:
        index + 1 === current
          ? fontsEnums[currentStatus]
          : index + 1 < current
          ? fontsEnums['DONE']
          : fontsEnums['WAIT'],
      stepStatus:
        index + 1 === current ? iconEnums[currentStatus] : index + 1 < current ? iconEnums['DONE'] : iconEnums['WAIT'],
      stepIconColor:
        index + 1 === current
          ? statusEnums[currentStatus]
          : index + 1 < current
          ? statusEnums['DONE']
          : statusEnums['WAIT']
    }))
  }

  return (
    <Layout.Vertical id="steps-progress" className={css.stepsProgress} spacing="large">
      <div className={css.progressBar}>
        <ProgressBar {...progressBarProps} className={css.progressHeight} />
      </div>
      <Layout.Vertical spacing="large">
        {stepsWithStatus && stepsWithStatus.length ? (
          stepsWithStatus.map(step => (
            <div key={`${step}`} className={css.stepWrap}>
              <Icon name={step.stepStatus} size={16} color={step.stepIconColor} margin={{ right: 'small' }} />
              <Container className={css.stepLabelContainer} style={{ fontWeight: step.fonts } as React.CSSProperties}>
                {step.label}
              </Container>
            </div>
          ))
        ) : (
          <Text>No Steps Available</Text>
        )}
      </Layout.Vertical>
    </Layout.Vertical>
  )
}
