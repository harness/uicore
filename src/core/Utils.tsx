/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Classes } from '@blueprintjs/core'
import copy from 'clipboard-copy'
import cx from 'classnames'
import React, { KeyboardEvent, MouseEvent } from 'react'
import { Popover } from '../components/Popover/Popover'
import { Color, Intent, OptionalTooltip } from '@harness/design-system'
import { FormikContext } from 'formik'
import { get, isPlainObject } from 'lodash-es'
import css from './Utils.css'

function stopEvent(event: MouseEvent | KeyboardEvent): void {
  event.stopPropagation()
  event.preventDefault()
}

const randomId = (): string => Math.random().toString(36).substring(2)

function getIntentColors(intent: Intent): { color: string; backgroundColor: string } {
  let color: Color = Color.WHITE
  let backgroundColor

  switch (intent) {
    case Intent.PRIMARY:
      backgroundColor = Color.PRIMARY_7
      break
    case Intent.SUCCESS:
      backgroundColor = Color.GREEN_500
      break
    case Intent.WARNING:
      backgroundColor = Color.YELLOW_500
      break
    case Intent.DANGER:
      backgroundColor = Color.RED_500
      break
    default:
      color = Color.GREY_500
      backgroundColor = Color.WHITE
      break
  }

  return { color, backgroundColor }
}

// Convert UICore named color into real CSS color
const getRealCSSColor = (color: Color): string =>
  `var(--${color // eslint-disable-line
    .match(/[A-Z][a-z]+|[0-9]+|[a-z]+/g)!
    .join('-')
    .toLowerCase()})`

interface WrapOptionalTooltipProps extends OptionalTooltip {
  children: JSX.Element
}

export function WrapOptionalTooltip({ tooltip, tooltipProps, children }: WrapOptionalTooltipProps): React.ReactElement {
  const isDark = tooltipProps?.isDark
  const content =
    typeof tooltip === 'string' ? (
      <div className={css.tooltipContainer} color={(isDark && 'white') || undefined}>
        {tooltip}
      </div>
    ) : (
      tooltip
    )

  return tooltip ? (
    <Popover
      boundary="viewport"
      position="top"
      interactionKind="hover"
      {...tooltipProps}
      popoverClassName={cx(isDark ? Classes.DARK : undefined, tooltipProps?.popoverClassName)}
      content={content || ''}>
      {children}
    </Popover>
  ) : (
    children
  )
}

export const errorCheck = (name: string, formik?: FormikContext<any>): boolean | '' | 0 | undefined =>
  ((get(formik?.touched, name) || (formik?.submitCount && formik?.submitCount > 0)) &&
    get(formik?.errors, name) &&
    !isPlainObject(get(formik?.errors, name))) as boolean

export function romanize(num: number, isLowerCase = false): string | number {
  if (isNaN(num)) return NaN
  const digits = String(+num).split('')
  const key = [
    '',
    'C',
    'CC',
    'CCC',
    'CD',
    'D',
    'DC',
    'DCC',
    'DCCC',
    'CM',
    '',
    'X',
    'XX',
    'XXX',
    'XL',
    'L',
    'LX',
    'LXX',
    'LXXX',
    'XC',
    '',
    'I',
    'II',
    'III',
    'IV',
    'V',
    'VI',
    'VII',
    'VIII',
    'IX'
  ]

  let roman = ''
  let i = 3
  while (i--) roman = (key[parseInt(digits.pop() || '') + i * 10] || '') + roman
  const returnStr = Array(+digits.join('') + 1).join('M') + roman
  return isLowerCase ? returnStr.toLowerCase() : returnStr
}

export const getSelectComponentPlaceholder = (plc?: string) => {
  // idea is to display placeholder like `- Select -`
  // This is as per the latest designs
  if (plc) {
    return `- ${plc} -`
  }
  return `- Select -`
}

// adopted from https://github.com/sindresorhus/escape-string-regexp v5.0.0
export function escapeStringRegexp(str: string): string {
  // Escape characters with special meaning either inside or outside character sets.
  // Use a simple backslash escape when it’s always valid, and a `\xnn` escape when the simpler form would be disallowed by Unicode patterns’ stricter grammar.
  return str.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d')
}

const Utils = {
  stopEvent,
  copy,
  randomId,
  getIntentColors,
  getRealCSSColor,
  WrapOptionalTooltip,
  romanize,
  getSelectComponentPlaceholder,
  escapeStringRegexp
}

export { Utils }
