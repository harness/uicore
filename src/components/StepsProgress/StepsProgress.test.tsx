/*
 * Copyright 2020 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { StepsProgress } from './StepsProgress'
import { Intent } from '@blueprintjs/core'

describe('Steps Progress Component', () => {
  test('should render with default steps', () => {
    const { container } = render(
      <StepsProgress
        steps={[
          'Checking Delegate',
          'Establising Connection with the Delegate and Kubernetes',
          'Checking for heartbeat',
          'Verifying Connection'
        ]}
        intent={Intent.DANGER}
        current={3}
        currentStatus={'ERROR'}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
