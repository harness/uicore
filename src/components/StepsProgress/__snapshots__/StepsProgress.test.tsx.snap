// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps Progress Component should render with default steps 1`] = `
<div>
  <div
    class="vertical layout-spacing-large main stepsProgress spacing spacing-large"
    id="steps-progress"
  >
    <div
      class="progressBar"
    >
      <div
        class="bp3-progress-bar bp3-intent-danger bp3-no-stripes progressHeight"
      >
        <div
          class="bp3-progress-meter"
          style="width: 75%;"
        />
      </div>
    </div>
    <div
      class="vertical layout-spacing-large main spacing spacing-large"
    >
      <div
        class="stepWrap"
      >
        <span
          class="bp3-icon bp3-icon-tick main color color-green500 margin margin-right margin-right-small"
          icon="tick"
        >
          <svg
            data-icon="tick"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              tick
            </desc>
            <path
              d="M14 3c-.28 0-.53.11-.71.29L6 10.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l8-8A1.003 1.003 0 0014 3z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="font main stepLabelContainer style style-fontWeight style-fontWeight-bold"
          style="font-weight: bold;"
        >
          Checking Delegate
        </div>
      </div>
      <div
        class="stepWrap"
      >
        <span
          class="bp3-icon bp3-icon-tick main color color-green500 margin margin-right margin-right-small"
          icon="tick"
        >
          <svg
            data-icon="tick"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              tick
            </desc>
            <path
              d="M14 3c-.28 0-.53.11-.71.29L6 10.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l8-8A1.003 1.003 0 0014 3z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="font main stepLabelContainer style style-fontWeight style-fontWeight-bold"
          style="font-weight: bold;"
        >
          Establising Connection with the Delegate and Kubernetes
        </div>
      </div>
      <div
        class="stepWrap"
      >
        <span
          class="bp3-icon bp3-icon-error main color color-red500 margin margin-right margin-right-small"
          icon="error"
        >
          <svg
            data-icon="error"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              error
            </desc>
            <path
              d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm1 13h-2v-2h2v2zm0-3h-2v-7h2v7z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="font main stepLabelContainer style style-fontWeight style-fontWeight-"
        >
          Checking for heartbeat
        </div>
      </div>
      <div
        class="stepWrap"
      >
        <span
          class="bp3-icon bp3-icon-blank main color color-yellow500 margin margin-right margin-right-small"
          icon="blank"
        >
          <svg
            data-icon="blank"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              blank
            </desc>
          </svg>
        </span>
        <div
          class="font main stepLabelContainer style style-fontWeight style-fontWeight-normal"
          style="font-weight: normal;"
        >
          Verifying Connection
        </div>
      </div>
    </div>
  </div>
</div>
`;
